### DISCUSSION TITLE
Custom validation not working with async

### DISCUSSION THREAD

**Discussion by popovicieduard** [adonisjs/core#1333] [Help] [answered]
When using async with a custom validation rule, the rule passes and doens't invoke the errorReport.report() function, if I remove the async from the rule, everything works as intended, I would like to use the Database inside my rule so I can check for a unique constraint on a table based on some values.

my rule:

```ts
validator.rule('checkUniqueAccountEmail',async (value, _:any,{ field, arrayExpressionPointer,errorReporter,root,tip})=>{
  /**
     * Skip validation when value is not a string. The string
     * schema rule will handle it 
     */
  if (typeof (value) !== 'string') {
    return
  }

  if (!value) {
    return
  }

  const accountId = validator.helpers.getFieldValue('accountId', root, tip) || null

  console.log(accountId)

  const userId = validator.helpers.getFieldValue('userId', root, tip)

  console.log(userId)

  const account = await Database
    .from('accounts')
    .where({emailAddress: value})
    .where({userId})
    .whereNot({uuid: accountId})
    .first()

  console.log(account)

  if (account) {
    console.log('aici')
    errorReporter.report(field, 'checkUniqueAccountEmail', 'email_address is not unique', arrayExpressionPointer)
    return
  }
})
```

**Comment by thetutlage** [comment, marked as answer]
You have to tell the validator in advance, that you are going to create an async validation by defining a third argument.

```ts
async function validate () {}

async function compile () {
  return {
    async: true,
  }
}

validator.rule('checkUniqueAccountEmail', validate, compile)
```

Check the AdonisJS unique rule https://github.com/adonisjs/lucid/blob/develop/src/Bindings/Validator.ts#L188

**Reply by zmaglica** [reply]
@thetutlage Can you please put this in documentation? I lost couple of hours debugging/solving the same problem.

**Reply by popovicieduard** [reply]
@zmaglica you can create a PR for docs. https://github.com/adonisjs/adonisjs.com
